BTrees/_base.py

```diff
@@ -815,9 +815,7 @@ def __delitem__(self, key):
         self._del(self._to_key(key))
 
     def clear(self):
-        if self._data:
-            # In the case of __init__, this was already set by __new__
-            self._data = []
+        self._data = []
```

Comment:
Maybe what I really don't like is how `clear` is called when we're in an entirely unknown state. Imagine a subclass of the BTree (ill-advised, but known to happen) that does something interesting in its clear method---maybe records some stats to graphite or something. If `clear` is called by `__setstate__` and during steady state, that subclass will now be broken, when it wasn't before.

- It's not safe for `__setstate__` to call arbitrary methods until after it's done.
- `__setstate__` is already coupled to `__getstate__` and the internal implementation of the object.

Maybe it would be better for `__setstate__` to simply assign to `self._data` *itself* and avoid the call to clear entirely. Then there's no object churn for the common case of `__init__` -> `clear`, and there's no implicit assumptions about the object state in `__setstate__`, it's all explicit. 

I think I like that best of all, so far.